src/detect-bytemath.c

```diff
@@ -0,0 +1,1364 @@
+                       "seen in byte_math - %s\n", rvalue);
+            goto error;
+        }
+        data->rvalue = index;
+        data->flags |= DETECT_BYTEMATH_RVALUE_VAR;
+        SCFree(rvalue);
+        rvalue = NULL;
+    }
+
+    SigMatch *prev_bmd_sm = DetectGetLastSMByListId(s, sm_list,
+            DETECT_BYTEMATH, -1);
+    if (prev_bmd_sm == NULL)
+        data->local_id = 0;
+    else
+        data->local_id = ((DetectByteMathData *)prev_bmd_sm->ctx)->local_id + 1;
+    if (data->local_id > de_ctx->byte_extract_max_local_id)
+        de_ctx->byte_extract_max_local_id = data->local_id;
+
+    sm = SigMatchAlloc();
+    if (sm == NULL)
+        goto error;
+    sm->type = DETECT_BYTEMATH;
+    sm->ctx = (void *)data;
+    SigMatchAppendSMToList(s, sm, sm_list);
+
+    if (!(data->flags & DETECT_BYTEMATH_FLAG_RELATIVE))
+        goto okay;
+
+    if (prev_pm == NULL)
+        goto okay;
+
+    if (prev_pm->type == DETECT_CONTENT) {
+        DetectContentData *cd = (DetectContentData *)prev_pm->ctx;
+        cd->flags |= DETECT_CONTENT_RELATIVE_NEXT;
+    } else if (prev_pm->type == DETECT_PCRE) {
+        DetectPcreData *pd = (DetectPcreData *)prev_pm->ctx;
+        pd->flags |= DETECT_PCRE_RELATIVE_NEXT;
+    }
+
+ okay:
+    return 0;
+
+ error:
+    if (rvalue)
+        SCFree(rvalue);
+    DetectByteMathFree(data);
+    return ret;
+}
+
+/**
+ * \brief Used to free instances of DetectByteMathractData.
+ *
+ * \param ptr Instance of DetectByteMathData to be freed.
+ */
+static void DetectByteMathFree(void *ptr)
+{
+    if (ptr != NULL) {
+        DetectByteMathData *bmd = ptr;
+        if (bmd->result != NULL)
+            SCFree((void *)bmd->result);
+        SCFree(bmd);
+    }
+
+    return;
+}
+
+/**
+ * \brief Lookup the SigMatch for a named byte_math variable.
+ *
+ * \param arg The name of the byte_math variable to lookup.
+ * \param s Pointer the signature to look in.
+ *
+ * \retval A pointer to the SigMatch if found, otherwise NULL.
+ */
+SigMatch *DetectByteMathRetrieveSMVar(const char *arg, const Signature *s)
+{
+    const int nlists = s->init_data->smlists_array_size;
+    for (int list = 0; list < nlists; list++) {
+        SigMatch *sm = s->init_data->smlists[list];
+        while (sm != NULL) {
+            if (sm->type == DETECT_BYTEMATH) {
+                const DetectByteMathData *bmd = (const DetectByteMathData *)sm->ctx;
+                if (strcmp(bmd->result, arg) == 0) {
+                    SCLogDebug("Retrieved SM for \"%s\"", arg);
+                    return sm;
+                }
+            }
+            sm = sm->next;
+        }
+    }
+
+    return NULL;
+}
+
+/*************************************Unittests********************************/
+#ifdef UNITTESTS
+
+#include "app-layer-parser.h"
+
+static int DetectByteMathParseTest01(void)
+{
+    int result;
+
+    DetectByteMathData *bmd = DetectByteMathParse("bytes 4, offset 2, oper +,"
```

Comment:
new tests should be written using `FAIL_IF`/`PASS` macros w/o any conditional statements.